logprep/framework/rule_tree/rule_parser.py

```diff
@@ -126,15 +130,36 @@ def _add_exists_filter(parsed_rules: list):
         for parsed_rule in parsed_rules:
             temp_parsed_rule = parsed_rule.copy()
             skipped_counter = 0
-
-            for segment_index, segment in enumerate(temp_parsed_rule):
-                if isinstance(segment, (Exists, Not, Always)):
+            for segment_idx, segment in enumerate(temp_parsed_rule):
+                if isinstance(segment, (Exists, Always)):
                     skipped_counter += 1
                     continue
 
-                exists_filter = Exists(segment.key)
-                if exists_filter in parsed_rule:
+                exists_filter = RuleParser._get_exists_filter(segment)
+                if exists_filter is None or exists_filter in parsed_rule:
                     skipped_counter += 1
-                    continue
+                else:
+                    parsed_rule.insert(segment_idx * 2 - skipped_counter, exists_filter)
```

Comment:
Can you explain where the 2 is coming from? I was trying to understand this and my first idea was that `segment_idx - 1` should work as well, but it doesn't. Maybe you could give this magic number a meaningful name?

Reply:
It was a way to account for the new elements added to the list.

`segment_idx * 2 - skipped_counter` 
<=> `segment_idx - (segment_idx  - skipped_counter)`
<=> `segment_idx - added_exists_filter_count`

It was indeed confusing, so I changed it.